File: src/schur_weyl/symmetric_functions.py
import numpy as np


def schur_polynomial(lam: tuple[int, ...], xs: list[float]) -> float:
    """Get the schur polynomial of a partition evaluated on input xs
    Not very precise above len(lam) = 7

    Args:
        lam (tuple[int, ...]): The partition
        xs (list[float]): The input variables

    Returns:
        float: the value of the polynomial
    """
    return _jacobi_trudi(lam, xs)


def _jacobi_trudi(lam: tuple[int, ...], xs: list[float]) -> float:
    """Not very precise above len(lam) = 7. Could use an other det formula"""
    if not lam:
        return 1.0
    ell: int = len(lam)
    h_k: list[float] = h_0_to_k(ell + lam[0] - 1, xs)
    mat = np.array(
        [[h_k[lam[i] - i + j] if lam[i] - i + j >= 0 else 0.0 for j in range(ell)] for i in range(ell)],
        dtype=float,
    )
    return float(np.linalg.det(mat))


def h_0_to_k(k: int, xs: list[float]) -> list[float]:
    """Completely homogenous symmetric polynomial from h_0 to h_k

    Args:
        k (int): k
        xs (list[float]): The input values

    Returns:
        list[float]: h_0 to h_k
    """
    h_i: list[float] = [1.0] + [0.0] * k
    for xi in xs:
        for i in range(1, k + 1):
            h_i[i] += xi * h_i[i - 1]
    return h_i


def power_sum(mu: tuple[int, ...], xs: list[float]) -> float:
    """p_mu(x) = prod_i (sum_j x_j^{mu_i}).

    Examples:
        >>> power_sum((1,), [2, 3])        # p_1 = 2 + 3
        5
        >>> power_sum((2,), [2, 3])        # p_2 = 4 + 9
        13
        >>> power_sum((2, 1), [2, 3])     # p_2 * p_1 = 13 * 5
        65
        >>> power_sum((), [2, 3])
        1
    """
    result = 1
    for part in mu:
        result *= sum(xj**part for xj in xs)
    return result

File: docs/src/schur-weyl/symmetricFunctions.ts
/**
 * Ported from `src/schur_weyl/symmetric_functions.py`.
 */

import type { Partition } from './youngDiagrams';
import { determinant } from './matrix';

/**
 * The Schur polynomial of a partition evaluated on input xs.
 * Not very precise above lam.length = 7 (see {@link determinant}).
 */
export function schurPolynomial(lam: Partition, xs: number[]): number {
  return jacobiTrudi(lam, xs);
}

/** Not very precise above lam.length = 7. Could use a different det formula. */
function jacobiTrudi(lam: Partition, xs: number[]): number {
  if (lam.length === 0) return 1.0;
  const ell = lam.length;
  const hk = h0ToK(ell + lam[0] - 1, xs);
  const mat: number[][] = [];
  for (let i = 0; i < ell; i++) {
    const row: number[] = [];
    for (let j = 0; j < ell; j++) {
      const idx = lam[i] - i + j;
      row.push(idx >= 0 ? hk[idx] : 0.0);
    }
    mat.push(row);
  }
  return determinant(mat);
}

/**
 * Completely homogeneous symmetric polynomials h_0 through h_k.
 *
 * @returns an array of length k + 1, [h_0, ..., h_k]
 */
export function h0ToK(k: number, xs: number[]): number[] {
  const h: number[] = [1.0, ...new Array<number>(k).fill(0.0)];
  for (const xi of xs) {
    for (let i = 1; i <= k; i++) {
      h[i] += xi * h[i - 1];
    }
  }
  return h;
}

/**
 * p_mu(x) = prod_i (sum_j x_j^{mu_i}).
 *
 * @example
 * powerSum([1], [2, 3]) // p_1 = 2 + 3 = 5
 * powerSum([2], [2, 3]) // p_2 = 4 + 9 = 13
 * powerSum([2, 1], [2, 3]) // p_2 * p_1 = 13 * 5 = 65
 * powerSum([], [2, 3]) // 1
 */
export function powerSum(mu: Partition, xs: number[]): number {
  let result = 1;
  for (const part of mu) {
    result *= xs.reduce((acc, xj) => acc + xj ** part, 0);
  }
  return result;
}
